File: lagom-persistence-couchbase/scaladsl/src/test/scala/com/lightbend/lagom/scaladsl/persistence/couchbase/TestEntityReadSide.scala
/*
 * Copyright (C) 2018 Lightbend Inc. <http://www.lightbend.com>
 */

package com.lightbend.lagom.scaladsl.persistence.couchbase

import akka.Done
import akka.actor.ActorSystem
import akka.persistence.couchbase.AsyncCouchbaseSession
import akka.stream.alpakka.couchbase.scaladsl.CouchbaseSession
import com.couchbase.client.java.document.JsonDocument
import com.couchbase.client.java.document.json.JsonObject
import com.lightbend.lagom.internal.persistence.couchbase.CouchbaseAction
import com.lightbend.lagom.scaladsl.persistence.ReadSideProcessor.ReadSideHandler
import com.lightbend.lagom.scaladsl.persistence.{AggregateEventTag, EventStreamElement, ReadSideProcessor, TestEntity}

import scala.collection.immutable
import scala.concurrent.{ExecutionContext, Future}

object TestEntityReadSide {
  class TestEntityReadSideProcessor(system: ActorSystem, readSide: CouchbaseReadSide)
      extends ReadSideProcessor[TestEntity.Evt] {

    def buildHandler(): ReadSideHandler[TestEntity.Evt] = {
      import system.dispatcher

      def updateCount(element: EventStreamElement[TestEntity.Appended]): Future[immutable.Seq[CouchbaseAction]] =
        Future.successful(immutable.Seq(new CouchbaseAction {
          override def execute(ab: CouchbaseSession, ex: ExecutionContext): Future[Done] =
            getCount(ab, element.entityId)
              .flatMap((count: Long) => {
                ab.upsert(
                  JsonDocument.create(s"count-${element.entityId}", JsonObject.create().put("count", count + 1))
                )
              })
              .map(_ => Done)
        }))

      readSide
        .builder[TestEntity.Evt]("testoffsets")
        .setEventHandler[TestEntity.Appended](updateCount)
        .build()
    }

    def aggregateTags: Set[AggregateEventTag[TestEntity.Evt]] = TestEntity.Evt.aggregateEventShards.allTags

  }

  def getCount(session: CouchbaseSession, entityId: String)(implicit ec: ExecutionContext): Future[Long] =
    session.get(s"count-$entityId").map {
      case Some(l) => l.content().getLong("count")
      case None => 0L
    }

}

class TestEntityReadSide(system: ActorSystem, couchbase: CouchbaseSession) {

  import system.dispatcher

  def getAppendCount(entityId: String): Future[Long] =
    TestEntityReadSide.getCount(couchbase, entityId)
}

File: lagom-persistence-couchbase/javadsl/src/test/java/com/lightbend/lagom/javadsl/persistence/TestEntityReadSide.java
/*
 * Copyright (C) 2018 Lightbend Inc. <http://www.lightbend.com>
 */

package com.lightbend.lagom.javadsl.persistence;

import akka.Done;
import akka.stream.alpakka.couchbase.javadsl.CouchbaseSession;
import com.couchbase.client.java.document.JsonDocument;
import com.couchbase.client.java.document.json.JsonObject;
import com.lightbend.lagom.javadsl.persistence.couchbase.CouchbaseReadSide;
import com.lightbend.lagom.javadsl.persistence.couchbase.CouchbaseAction;
import org.pcollections.PSequence;

import javax.inject.Inject;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

public class TestEntityReadSide {

  private final CouchbaseSession session;

  @Inject
  public TestEntityReadSide(CouchbaseSession session) {
    this.session = session;
  }

  public CompletionStage<Long> getAppendCount(String entityId) {
    return getCount(session, entityId);
  }

  private static CompletionStage<Long> getCount(CouchbaseSession session, String entityId) {
    return session.get("count-" + entityId).thenApply(v ->
        v.isPresent() ? v.get().content().getLong("count") : 0
    );
  }

  public static class TestEntityReadSideProcessor extends ReadSideProcessor<TestEntity.Evt> {
    private final CouchbaseReadSide readSide;

    @Inject
    public TestEntityReadSideProcessor(CouchbaseReadSide readSide) {
      this.readSide = readSide;
    }

    @Override
    public ReadSideProcessor.ReadSideHandler<TestEntity.Evt> buildHandler() {
      return readSide.<TestEntity.Evt>builder("testoffsets")
          .setEventHandler(TestEntity.Appended.class, this::updateCount)
          .build();
    }

    private CompletionStage<List<CouchbaseAction>> updateCount(TestEntity.Appended event) {
      return CompletableFuture.completedFuture(Collections.singletonList(session ->
          getCount(session, event.getEntityId()).thenComposeAsync(count -> {
            JsonObject content = JsonObject.create().put("count", count + 1);

            return session
                .upsert(JsonDocument.create("count-" + event.getEntityId(), content))
                .thenApply(d -> Done.getInstance());
          })
      ));
    }

    @Override
    public PSequence<AggregateEventTag<TestEntity.Evt>> aggregateTags() {
      return TestEntity.Evt.AGGREGATE_EVENT_SHARDS.allTags();
    }
  }

}
